rfcs/react-components/convergence/theme-shared-colors.md

```diff
@@ -0,0 +1,114 @@
+
+Just adding v9 `Provider` with a theme, the test time rises by **20ms** (with the original approach having 1200 tokens and alias references it was 80ms) which does not pass the gate. For that reason we have been stuck on this for 3 months already not being able to enable the v9 `Provider` in the application, blocking feature teams in starting migration to v9.
+
+As an experiment we tried to reduce the tokens to 200, the test time degradation was 7ms.
+
+| Theme structure               | WLT degradation [ms] |
+| ----------------------------- | -------------------: |
+| 1200 tokens, alias references |                   70 |
+| 600 tokens                    |                   20 |
+| 200 tokens                    |                    7 |
+
+**The numbers confirm the performance decreases linearly depending on the number of CSS variables (and therefore theme tokens).**
+
+### New Shared Color Alias tokens request
+
+_The numbers might be inaccurate as the feature is not fully spec-ed yet!_
+
+In order to support new scenarios (colorful calendar view), there is a new request to add 3 new shared color alias tokens. Adding them would result in adding 3 alias tokens for each of 49 shared colors = adding 147 new shared color alias tokens which would affect application performance.
+
+## Detailed Design or Proposal
+
+The theme currently contains 49 global shared colors.
+
+Current alias structure forces each ramp to be uniform (the same set of alias tokens for each shared color), resulting in 49 alias tokens each time we need a new line item (new alias token). Today we have 9 alias slots = total alias count is 441 (49x9).
+
+Components have varying degrees of color needs.
+
+Instead of 1 uniform Shared alias table (today's experience), separate out into 3 tables:
+
+1. Semantic States
+   - There are 5 unique global colors needed to support this set (Danger, Success, Severe, Warning, Out of office)
+   - Assume 9 alias tokens (Could potentially remove 2 from this table as they were avatar specific)
+   - 45 alias tokens in total (5 x 9)
+2. Persona
+   - There are 30 unique global colors needed to support this set
+   - Only needs 3 alias tokens
+   - 90 tokens in total (30 x 3)
+3. Category (calendar categories) (not fully defined yet)
+   - 10 unique global colors to support this set
+   - Assuming 9 alias slots (subject to increase)
+   - 90 tokens in total (10 x 9)
+
+Dividing up by 3 tables would reduce token count from 441 to ~225 (Table 1+Table 2+Table 3) while adding support for Calendar scenario.
+
+### Pros and Cons
+
+- 👍 Reducing shared color alias tokens from 441 to ~225 instead of increasing them to ~600 (to add tokens needed for Calendar)
+- 👍 More clarity around which specific color to use for semantic states instead of relying on design guidance
+- 👍 Allows a bit more wiggle room for the semantic alias table to grow w/o adding too much bloat
+- 👎 **Breaking change after the final RC**
```

Comment:
For me this is the reason to not make this change, or rather, to make it in a non-breaking way.

As you note below this should not affect partners much but my experience with other breaking changes is that they can require quite a bit of coordination between partners. Historically we've reserved the right to make breaking changes to resolve accessibility issues and my belief is that performance is an accessibility issue but we should still strive to make these types of changes in a non-breaking way whenever possible.

While it is not ideal to ship a first release with deprecated features (in fact this was a motivation for breaking changes we've previously made during RC) my feeling is that we're getting too close to the stable release to ship a breaking change that isn't resolving a hard blocking issue. I don't feel this is a hard blocking issue for release as we can ship "v2" or "perf" versions of all themes and update our docs to demonstrate their use in all cases.

I do think this is a good change to make and I'm not strictly opposed to making it in a breaking way, I would just prefer to do so in a backwards compatible way given our [release schedule](https://github.com/microsoft/fluentui/wiki/Fluent-UI-React-v9-Release).

Reply:
I tend to agree with @spmonahan here, we should do this either in a non-breaking way or change our release schedule.

Reply:
Can we please prioritize actually asking partners about this? they stand the most to gain/lose for perf/breaking change in this scenario

Reply:
> I do think this is a good change to make and I'm not strictly opposed to making it in a breaking way, I would just prefer to do so in a backwards compatible way given our [release schedule](https://github.com/microsoft/fluentui/wiki/Fluent-UI-React-v9-Release).

we agreed that BC in RC phase can be done if really necessary, so landing this prior to stable should be ok. 

Also dunno if you folks read reasoning regarding why v2 was discarded, but it gives you good additional context

Reply:
@spmonahan has been working most closely with partners on BC, so he should have a sense of their tolerance. That being said, I agree with @ling1726 that flighting this with our core partners is a the right call.

Reply:
let me put on my partner hat. v9 adoption in Teams is currently blocked by the perf issue - injecting too many tokens delays initial render and degreades some of the perf metrics. I don't know if all partners track this metrics but probably the initial render is critical for most of them. In production rings this is up to 300ms regression in average and we have seen the cost being linear with relation to the amount of tokens.

Migration to new theme might have some cost for the partners and it will have cost for Teams, but as far as I undrstand it, it will be a change that breaks build and is easily detectable.

If we decide to deprecate the old theme, partners will need to migrate to the new theme anyway.

Since we are in RC, I think we should not release final version knowing that a change like this is planned. We should make the change and explain the benefits.